torch/_inductor/lowering.py

```diff
@@ -3358,6 +3358,23 @@ def _unsafe_index_put_(self, indices, values, accumulate=False):
 
 
 def index_put_impl_(self, indices, values, accumulate, check):
+    def try_get_name(x):
+        if isinstance(x, ir.TensorBox):
+            x = x.data
+        if isinstance(x, ir.BaseView):
+            x = x.unwrap_view()
+        if isinstance(x, ir.StorageBox):
+            x = x.data
+        return x.get_name() if isinstance(x, ir.Buffer) else None
+
+    if try_get_name(self) in values.get_read_names():
+        # Fix issue: https://github.com/pytorch/pytorch/issues/138908
+        # When self and values have memory overlapping, indices may
+        # contain duplicate values, potentially causing incorrect results since
+        # the load of `values` might contain modified value from the store of `self`.
+        # To address this, store values in a temporary buffer in such cases.
+        values.realize()
```

Comment:
Could we only realize this when values and self alias ?

Reply:
I believe we are checking the aliasing between `self` and `values` by verifying if the buffer name of `self` appears in any read buffers of `values`. Is this the same as what you mentioned about checking the aliasing of `values` and `self`?

Reply:
Yea, this looks good. however the benchmakrs seemingly show a compile time regression. i believe this is because `get_read_names()` calls `inner_fn_opcount` which isnt cheap.

Can we skip this check for the `index_put` and `_unsafe_index_put` lowerings ? in this case `self` is freshly constructed and we dont need to check.